library/std/src/thread/mod.rs

```diff
@@ -1076,6 +1078,27 @@ impl ThreadId {
         }
     }
 
+    #[cfg(target_has_atomic = "64")]
+    fn new() -> ThreadId {
+        use crate::sync::atomic::AtomicU64;
+
+        static COUNTER: AtomicU64 = AtomicU64::new(1);
+        const COUNTER_MAX: u64 = 1u64 << 63;
+
+        let id = COUNTER.fetch_add(1, SeqCst);
+
+        /* We need to have a safe margin so that if many threads call fetch_add
+         * we won't accidentally overflow and return potentially non-unique IDs */
+        if id == COUNTER_MAX {
+            COUNTER.store(COUNTER_MAX, SeqCst);
+        }
+        if id >= COUNTER_MAX {
```

Comment:
```suggestion
        if id >= COUNTER_MAX {
            COUNTER.store(COUNTER_MAX, SeqCst);
```

To avoid race on overflow, since another thread might have read value above COUNTER_MAX already.

Reply:
Well, if it reads value above COUNTER_MAX, it means some other thread read COUNTER_MAX and that thread is responsible for storing COUNTER_MAX back

Reply:
Ah, indeed there is no need for other threads to store it back.